### PR TITLE
Basic navigator unit tests

### PR DESCRIPTION
**Pull request by bgrayd**
These are unit tests for the basic functionality of the navigator and also works as an example unit test for others to use when making a new unit test.  It also includes moving some of the functions that do not need to be part of the navigator node class outside of the class, so that they can be tested without ROS.

### CODE REVIEW

File: cr17/test/test_navigator.py [hunk 1]
```diff
@@ -1,17 +1,121 @@
...
+
 
 
 class TestNavigator(unittest.TestCase):
+    def assertListAlmostEqual(self, list1, list2, tol = 7):
+        self.assertEqual(len(list1), len(list2))
+        for a, b in zip(list1, list2):
+             self.assertAlmostEqual(a, b, tol)
+
+    def test_at_goal(self):
```
Comment (JacobAMason) [inline cr17/test/test_navigator.py:20]: This test should include an assertion to test the tolerance feature of this function.

File: cr17/nodes/cr17/navigator.py [hunk 2]
```diff
@@ -27,6 +27,81 @@
...
+        d_x = 0
+        d_y = 0
+    elif (GOAL_THRESH <= dist) and (dist <= FIELD_SPREAD + GOAL_THRESH):
+        d_x = ALPHA * (dist - GOAL_THRESH) * math.cos(force_angle)
+        d_y = ALPHA * (dist - GOAL_THRESH) * math.sin(force_angle)
+    else: #dist > (FIELD_SPREAD + GOAL_THRESH)
+        d_x = ALPHA * FIELD_SPREAD * math.cos(force_angle)
+        d_y = ALPHA * FIELD_SPREAD * math.sin(force_angle)
+
+    return [d_x, d_y]
```
Comment (JacobAMason) [inline cr17/nodes/cr17/navigator.py]: Is this meant to be a list/mutable? Would it be better formulated as a tuple?

File: cr17/nodes/cr17/navigator.py [hunk 3]
```diff
@@ -27,6 +27,81 @@
...
+        
+    while angle <= -math.pi:
+        angle = angle + 2 * math.pi
+    return angle
+
+def calc_repulsive_force(obstacles, robot_pose):
+    '''
+    Given a list of obstacles and robot pose, calculate repulsive force
+    '''
+    rep_force = [0, 0]
```
Comment (JacobAMason) [inline cr17/nodes/cr17/navigator.py]: list/tuple

File: cr17/test/test_navigator.py [hunk 4]
```diff
@@ -1,17 +1,121 @@
...
+        robotPose.position.y = 0
+
+        self.assertListAlmostEqual([0.9,0], navigator.calc_goal_force(goalForward, robotPose))
+        self.assertListAlmostEqual([-0.9,0], navigator.calc_goal_force(goalBackward, robotPose))
+        self.assertListAlmostEqual([0,0.9], navigator.calc_goal_force(goalLeft, robotPose))
+        self.assertListAlmostEqual([0,-0.9], navigator.calc_goal_force(goalRight, robotPose))
+        self.assertListAlmostEqual([0.929,0.929], navigator.calc_goal_force(goalFrontLeft, robotPose), 3)
+        self.assertListAlmostEqual([0.929,-0.929], navigator.calc_goal_force(goalFrontRight, robotPose), 3)
+        self.assertListAlmostEqual([-0.929,0.929], navigator.calc_goal_force(goalBackLeft, robotPose), 3)
+        self.assertListAlmostEqual([-0.929,-0.929], navigator.calc_goal_force(goalBackRight, robotPose), 3)
```
Comment (JacobAMason) [inline cr17/test/test_navigator.py]: list/tuple

File: cr17/test/test_navigator.py [hunk 5]
```diff
@@ -1,17 +1,121 @@
...
+
+    def test_calc_repulsive_force(self):
+        obstacle1 = [0,0.4]
+        obstacle2 = [0,-0.4]
+        obstacle3 = [0.4,0]
+        obstacle4 = [-0.4,0]
```
Comment (JacobAMason) [inline cr17/test/test_navigator.py]: list/tuple

File: cr17/test/test_navigator.py [hunk 6]
```diff
@@ -1,17 +1,121 @@
...
+        
+        self.assertListAlmostEqual([-0.11,-0.11], navigator.calc_repulsive_force([obstacle3, obstacle1],robotPose))
+        self.assertListAlmostEqual([-0.11, 0.11], navigator.calc_repulsive_force([obstacle3, obstacle2],robotPose))
+        self.assertListAlmostEqual([-0.22, 0], navigator.calc_repulsive_force([obstacle3, obstacle3],robotPose))
+        self.assertListAlmostEqual([0,0], navigator.calc_repulsive_force([obstacle3, obstacle4],robotPose))
+
+        self.assertListAlmostEqual([0.11,-0.11], navigator.calc_repulsive_force([obstacle4, obstacle1],robotPose))
+        self.assertListAlmostEqual([0.11, 0.11], navigator.calc_repulsive_force([obstacle4, obstacle2],robotPose))
+        self.assertListAlmostEqual([0,0], navigator.calc_repulsive_force([obstacle4, obstacle3],robotPose))
+        self.assertListAlmostEqual([0.22, 0], navigator.calc_repulsive_force([obstacle4, obstacle4],robotPose))
```
Comment (JacobAMason) [inline cr17/test/test_navigator.py]: list/tuple

**Review by JacobAMason — CHANGES_REQUESTED** [review]

File: cr17/test/test_navigator.py (hunk 1 above)
Comment (bgrayd) [reply cr17/test/test_navigator.py:20]: I believe this was fixed in 79e09e4a7334979cb3e82fb2bf824e29fd1342ad

File: cr17/test/test_navigator.py (hunk 4 above)
Comment (bgrayd) [reply cr17/test/test_navigator.py]: Fixed in 2b4de4372834138b7cde16f7b3bd395bf3d75540

File: cr17/test/test_navigator.py (hunk 5 above)
Comment (bgrayd) [reply cr17/test/test_navigator.py]: Fixed in 2b4de4372834138b7cde16f7b3bd395bf3d75540

File: cr17/test/test_navigator.py (hunk 6 above)
Comment (bgrayd) [reply cr17/test/test_navigator.py]: Fixed in 2b4de4372834138b7cde16f7b3bd395bf3d75540

File: cr17/nodes/cr17/navigator.py (hunk 2 above)
Comment (bgrayd) [reply cr17/nodes/cr17/navigator.py]: There was a place in a different function where it normalized the vector, but that was changed to use tuples. Changed in 2b4de4372834138b7cde16f7b3bd395bf3d75540

File: cr17/nodes/cr17/navigator.py (hunk 3 above)
Comment (bgrayd) [reply cr17/nodes/cr17/navigator.py]: Fixed in 2b4de4372834138b7cde16f7b3bd395bf3d75540

File: cr17/nodes/cr17/navigator.py [hunk 7]
```diff
@@ -97,8 +97,7 @@ def calc_repulsive_force(obstacles, robot_pose):
             # obstacle is far away, don't care about it
             d_x = 0
             d_y = 0
-        rep_force[0] += d_x
-        rep_force[1] += d_y
+        rep_force = (rep_force[0]+d_x, rep_force[1]+d_y)
```
Comment (JacobAMason) [inline cr17/nodes/cr17/navigator.py]: might want to leave this as a list and then `return tuple(rep_force)`

**Review by JacobAMason — APPROVED** [review]

File: cr17/test/test_navigator.py (hunk 1 above)
Comment (JacobAMason) [reply cr17/test/test_navigator.py:20]: It is indeed

**Pull request merged**